tbtc-ts/src/sweep.ts

```diff
@@ -0,0 +1,372 @@
+        depositData,
+        previousUtxoWithRaw
+      )
+    : await createSweepTransaction(
+        fee,
+        walletPrivateKey,
+        utxosWithRaw,
+        depositData
+      )
+
+  // TODO: `broadcast` may fail silently (i.e. no error will be returned, even
+  // if the transaction is rejected by other nodes and does not enter the
+  // mempool, for example due to an UTXO being already spent).
+  // We could call `verify` on the transaction before broadcasting which would
+  // ensure it is correctly constructed, but this check alone does not ensure
+  // it's correct from the network point of view (for example the UTXOs must all
+  // be spendable).
+  // One solution could be checking the mempool after broadcast or simply do
+  // nothing and assume the caller used correct UTXOs.
+  await bitcoinClient.broadcast(transaction)
+}
+
+/**
+ * Creates a Bitcoin P2WPKH sweep transaction.
+ * @param fee - the value that should be subtracted from the sum of the UTXOs
+ *              values and used as the transaction fee.
+ * @param walletPrivateKey - Bitcoin private key of the wallet.
+ * @param utxos - UTXOs from new deposit transactions. Must be P2(W)SH.
+ * @param depositData - data on deposits. Each elements corresponds to UTXO. The
+ *                      number of UTXOs and deposit data elements must equal.
+ * @param previousUtxo - UTXO from the previous sweep transaction (optional).
+ *                       Must be P2WPKH.
+ * @returns Bitcoin sweep transaction in raw format.
+ */
+export async function createSweepTransaction(
+  fee: BigNumber,
+  walletPrivateKey: string,
+  utxos: (UnspentTransactionOutput & RawTransaction)[],
+  depositData: DepositData[],
+  previousUtxo?: UnspentTransactionOutput & RawTransaction
+): Promise<RawTransaction> {
+  if (utxos.length < 1) {
+    throw new Error("There must be at least one deposit UTXO to sweep")
+  }
+  const decodedWalletPrivateKey = wif.decode(walletPrivateKey)
+
+  const walletKeyRing = new bcoin.KeyRing({
+    witness: true,
+    privateKey: decodedWalletPrivateKey.privateKey,
+    compressed: decodedWalletPrivateKey.compressed,
+  })
+
+  const walletAddress = walletKeyRing.getAddress("string")
+
+  const inputCoins = []
+  let totalInputValue = 0
+
+  if (previousUtxo) {
+    inputCoins.push(
+      bcoin.Coin.fromTX(
+        bcoin.MTX.fromRaw(previousUtxo.transactionHex, "hex"),
+        previousUtxo.outputIndex,
+        -1
+      )
+    )
+    totalInputValue += previousUtxo.value
+  }
+
+  for (const utxo of utxos) {
+    inputCoins.push(
+      bcoin.Coin.fromTX(
+        bcoin.MTX.fromRaw(utxo.transactionHex, "hex"),
+        utxo.outputIndex,
+        -1
+      )
+    )
+    totalInputValue += utxo.value
+  }
+
+  const transaction = new bcoin.MTX()
+
+  transaction.addOutput({
+    script: bcoin.Script.fromAddress(walletAddress),
+    value: totalInputValue,
+  })
+
+  await transaction.fund(inputCoins, {
+    changeAddress: walletAddress,
```

Comment:
`fund` requires the caller to provide `change Address`:
```
assert(options.changeAddress, 'Change address is required.');
```

I thought about writing a custom `fund` function, but the `fund` from `bcoin` works fine in our case.   

In our case, there won't be any change output created since we request just one output for the sweep transaction with `value` set to exactly  the sum of all the inputs.

Do you suggest adding an assertion there is only one output created, just in case?

Reply:
Yep, I think an assertion about single output will do the trick.

Reply:
Done in https://github.com/keep-network/tbtc-v2/pull/96/commits/24c04ee98e13f36b221bbb8f2d0467422be4130b.